### PR TITLE
Added api3 oracle contract

### PR DESCRIPTION
**Pull request by yogesh0509**
This is a work-in-progress pull request!!

I have added `readData()` function but for it to work we would also need to call `setDapiName()` function. 
Should I set it under `acceptTermsOfService()` function or create a different function for this task.

### CODE REVIEW

**Comment by ceilican** [comment]
Thanks for the WIP PR, @yogesh0509 ! 

> I have added readData() function but for it to work we would also need to call setDapiName() function.
Should I set it under acceptTermsOfService() function or create a different function for this task.

I am not completely sure. But:

* `acceptTermsOfService` is about accepting the terms of service of the oracle. As far as I know, API3 oracles do not require accepting any terms of service. Therefore, our `api3Oracle` contract should have an `acceptTermsOfService` that does nothing.

* setting the `dapiName` appears to be something that we need to do only once when setting up the API3 oracle reader. Therefore, I think it would be reasonable to pass the `dapiName` as an argument to the constructor of our `api3Oracle` and do any calls related to setting it up there in the body of the constructor function.

But I am not familiarized with API3 oracles and I don't know exactly what `dapiName` means or how it should be configured. Therefore, I am not certain of my comments above.

What do you think?

File: src/api3Oracle.sol [hunk 1]
```diff
@@ -0,0 +1,29 @@
+// SPDX-License-Identifier: AEL
+pragma solidity ^0.8.0;
+
+import "airnode-protocol-v1/api3-server-v1/DapiServer.sol";
+
+contract api3Oracle is DapiServer {
```
Comment (ceilican) [inline src/api3Oracle.sol:6]: Let's follow the Solidity naming conventions and rename this to `API3Oracle`.

File: src/api3Oracle.sol [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+        bytes32 _dapiNameHash
+    ) DapiServer(_accessControlRegistry, _adminRoleDescription, _manager) {
+        dapiNameHash = _dapiNameHash;
+    }
+
+    function readData() external view returns (uint256) {
+        int224 value;
+        uint32 timestamp;
+
+        (value, timestamp) = _readDataFeedWithDapiNameHash(dapiNameHash);
```
Comment (ceilican) [inline src/api3Oracle.sol:22]: We can remove lines 19 and 20 and write `(int224 value, uint32 timestamp) = ...` here instead.

File: src/api3Oracle.sol [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+        int256 signedValue = int256(value);
+
+        require(signedValue >= 0, "Cannot convert negative value");
```
Comment (ceilican) [inline src/api3Oracle.sol:25]: I think we can optimize the code by checking that `value >= 0` and then converting directly from `int224` to `uint256`, without first converting to `int256` and assigning a new local variable. Could you check this?

Comment (yogesh0509) [reply src/api3Oracle.sol:23]: We cannot directly convert an int224 to uint256 and need to convert int224 to int256 first.

File: src/api3Oracle.sol (hunk 2 above)
Comment (yogesh0509) [reply src/api3Oracle.sol]: Pushed the code with the changes suggested.

File: src/api3Oracle.sol (hunk 1 above)
Comment (yogesh0509) [reply src/api3Oracle.sol]: Pushed the code after renaming the file

**Comment by yogesh0509** [comment]
> Thanks for the WIP PR, @yogesh0509 !
> 
> > I have added readData() function but for it to work we would also need to call setDapiName() function.
> > Should I set it under acceptTermsOfService() function or create a different function for this task.
> 
> I am not completely sure. But:
> 
> * `acceptTermsOfService` is about accepting the terms of service of the oracle. As far as I know, API3 oracles do not require accepting any terms of service. Therefore, our `api3Oracle` contract should have an `acceptTermsOfService` that does nothing.
> * setting the `dapiName` appears to be something that we need to do only once when setting up the API3 oracle reader. Therefore, I think it would be reasonable to pass the `dapiName` as an argument to the constructor of our `api3Oracle` and do any calls related to setting it up there in the body of the constructor function.
> 
> But I am not familiarized with API3 oracles and I don't know exactly what `dapiName` means or how it should be configured. Therefore, I am not certain of my comments above.
> 
> What do you think?

* `dapiName` will be different for each deployment, so we can pass it as an argument in the constructor and set it up in the there itself. 
* For our milkomeda testnet, the `dapiName` will be **ADA/USD** as referenced [here](https://github.com/DjedAlliance/api3-oracle-data-feed-reader/tree/milkomeda-testnet)

File: src/api3Oracle.sol (hunk 3 above)
Comment (ceilican) [reply src/api3Oracle.sol:23]: OK. Thanks for letting me know!

Comment (ceilican) [reply src/api3Oracle.sol:23]: In this case, we can probably still optimize the code by:

1) doing `require(value >= 0, ...)`

and then 

2) `return uint256(int256(value))`

Can't we?

**Comment by ceilican** [comment]
> dapiName will be different for each deployment, so we can pass it as an argument in the constructor and set it up in the there itself.

OK. Let's do it like this.

> For our milkomeda testnet, the dapiName will be ADA/USD as referenced [here](https://github.com/DjedAlliance/api3-oracle-data-feed-reader/tree/milkomeda-testnet)

OK. It is unclear when API3 will deploy to the milkomeda mainnet though.

I was thinking we could consider having a deployment of Djed on Polygon using this oracle adapter to an API3 oracle on Polygon. Could you have a look on what the dapiName on Polygon would be? Would it be "MATIC/USD"? Would the adapter work on Polygon as it is? Or would it need changes?

File: src/api3Oracle.sol (hunk 3 above)
Comment (yogesh0509) [reply src/api3Oracle.sol:23]: Yes, it can be optimized this way. I have made the change and will push it with other changes.

**Comment by yogesh0509** [comment]
> OK. Let's do it like this.
> 
> >

The function `setDapiName` is an external function in the `DapiServer.sol` file which means we cannot set it up in the constructor itself. Maybe we can try calling this function in the deploy script.

**Comment by yogesh0509** [comment]
I looked into the docs and I have found another way to do the same. We don't need to set up a `DapiServer` and can use a proxy instead to fetch data from existing dAPI. 
Also, we already have a dAPI deployed on polygon with the dAPI name "MATIC/USD".

I will close this pr and open a new wip PR with the changes!!

**Pull request closed without merging**